src/main/java/randoop/output/FailingTestFilter.java

```diff
@@ -0,0 +1,277 @@
+
+    /*
+     * Split Java code text so that we can match the line number for the assertion with the code.
+     * Use same line break as used to write test class file.
+     */
+    String[] javaCodeLines = javaCode.split(Globals.lineSep);
+
+    int failureCount = 0;
+    while (failureCount < totalFailures) {
+      /*
+       * Read until beginning of failure
+       */
+      Pair<String, String> failureHeaderMatch = readUntilMatch(lineIterator, failureHeaderPattern);
+      String line = failureHeaderMatch.a;
+      String methodName = failureHeaderMatch.b;
+
+      /*
+       * If the method name in the failure message is not a test method, throw an exception.
+       */
+      if (!methodName.matches(GenTests.TEST_METHOD_NAME_PREFIX + "\\d+")) {
+        if (line.contains("initializationError")) {
+          throw new BugInRandoopException(
+              "Check configuration of test environment: "
+                  + "initialization error of test in flaky-test filter");
+        } else {
+          throw new BugInRandoopException("Unexpected failure in flaky-test filter: " + methodName);
+        }
+      }
+
+      /*
+       * Search for the stacktrace entry corresponding to the test method, and capture the line
+       * number.
+       */
+      Pattern linePattern =
+          Pattern.compile(
+              String.format(
+                  "\\s+at\\s+%s\\.%s\\(%s\\.java:(\\d+)\\)", classname, methodName, classname));
+
+      Pair<String, String> failureLineMatch = readUntilMatch(lineIterator, linePattern);
+      int lineNumber = Integer.parseInt(failureLineMatch.b);
+      if (lineNumber < 1 && lineNumber < javaCodeLines.length + 1) {
+        throw new BugInRandoopException(
+            "Line number "
+                + lineNumber
+                + " read from JUnit out of range [1,"
+                + (javaCodeLines.length + 1)
+                + "]");
+      }
+      javaCodeLines[lineNumber - 1] = "// flaky: " + javaCodeLines[lineNumber - 1];
+
+      failureCount++;
+    }
+
+    //XXX have this method return the array and redo writeClass so that it writes from array (?)
+    return UtilMDE.join(javaCodeLines, Globals.lineSep);
+  }
+
+  /**
+   * Read lines of the JUnit output using the iterator until finding a match for the pattern, and
+   * then returns a pair containing the line and the text matching the first group of the pattern.
+   * Assumes that there is a match, and that the pattern has at least one group.
+   *
+   * @param lineIterator the iterator for reading from the JUnit output
+   * @param pattern the pattern for a regex with at least one group
+   * @return the pair containing the line and the text matching the first group
+   * @throws BugInRandoopException if the iterator has no more lines, but the pattern hasn't been
+   *     matched
+   */
+  private Pair<String, String> readUntilMatch(Iterator<String> lineIterator, Pattern pattern) {
```

Comment:
The use of fields `a` and `b` makes client code *very* hard to read.
Create a new struct class (it just has to have two public fields with clear names) and use it instead of `Pair`.